### PR TITLE
[core][rfc] Fix race condition between write chunk and abort object.

### LINKED ISSUE #17080: [placeholder] Investigate Dask on Ray segfault from dlmalloc in nightly release tests
https://buildkite.com/ray-project/periodic-ci/builds/555#_

### PR DESCRIPTION
**Pull request by scv119**
TL;DR: there is a race condition between writing a chunk (of the object) vs abort the object creation. This could lead to memory corruption when an object is first removed(aborted) then the address being written.

We still seeing low frequency segfault in stress tests after fixing few issues: #17204, #17140. 
By examining the core dump, we were able to confirm that mmaped addesses in dlmalloc are corrupted, as SIGSEGV happens at [dlmalloc/unlink_chunk](https://github.com/ray-project/ray/blob/35ec91c4e04c67adc7123aa8461cf50923a316b4/src/ray/thirdparty/dlmalloc.c#L4729).

Given dlmalloc itself is very stable and we don't double-free , the only possible cause for this corruption is that we write to unallocated address. The current implementation of handlePush in object manager could have cause this. as chunks creating and abortion are handled by multiple threads, it could happen that one thread aborted the object creation where another thread still trying to write to that aborted chunk.  

The proposed solution in this PR is to synchronize the object abortion with chunk mutation by a mutex. 
If we think this is a bottleneck, we could create a lock for each object or shard locks by object_id.

```
187386 [2021-07-20 21:42:05,924 E 4550 4575] logging.cc:299: *** SIGSEGV received at time=1626817325 on cpu 3 ***
187387 [2021-07-20 21:42:05,924 E 4550 4575] logging.cc:299: PC: @     0x5614428e5b62  (unknown)  dlfree
187388 [2021-07-20 21:42:05,924 E 4550 4575] logging.cc:299:     @     0x561442d6927b         64  absl::lts_20210324::WriteFailureInfo()
187389 [2021-07-20 21:42:05,936 E 4550 4575] logging.cc:299:     @     0x561442d69452         96  absl::lts_20210324::AbslFailureSignalHandler()
187390 [2021-07-20 21:42:05,937 E 4550 4575] logging.cc:299:     @     0x7f21249e5890       1792  (unknown)
187391 [2021-07-20 21:42:05,937 E 4550 4575] logging.cc:299:     @     0x5614428b499c         80  plasma::PlasmaAllocator::Free()
187392 [2021-07-20 21:42:05,937 E 4550 4575] logging.cc:299:     @     0x5614428b92f9        112  plasma::PlasmaStore::EraseFromObjectTable()
```
```
(gdb) info line* 0x5614428e5b62
Line 4729 of "bazel-out/k8-dbg/bin/_virtual_includes/plasma_store_server_lib/ray/thirdparty/dlmalloc.c" starts at address 0x5614428e5870 <plasma::dlfree(void*)+448>
   and ends at 0x5614428e5f01 <plasma::dlfree(void*)+2129>.
(gdb) info line* 0x7f21249e5890
Line 28 of "../sysdeps/pthread/flockfile.c" starts at address 0x7f21249e5881 <__flockfile+17> and ends at 0x7f21249e58c2 <__flockfile+82>.
(gdb) info line* 0x5614428b499c
Line 82 of "src/ray/object_manager/plasma/plasma_allocator.cc" starts at address 0x5614428b499c <plasma::PlasmaAllocator::Free(void*, unsigned long)+154>
   and ends at 0x5614428b49c0 <plasma::PlasmaAllocator::Free(void*, unsigned long)+190>.
```

Test plan:
- [x] existing unit tests
- [x] stress test it.

### CODE REVIEW

File: src/ray/object_manager/object_buffer_pool.cc [hunk 1]
```diff
@@ -144,16 +144,22 @@ void ObjectBufferPool::AbortCreateChunk(const ObjectID &object_id,
...
   std::lock_guard<std::mutex> lock(pool_mutex_);
   auto it = create_buffer_state_.find(object_id);
   if (it == create_buffer_state_.end() ||
-      it->second.chunk_state[chunk_index] != CreateChunkState::REFERENCED) {
+      it->second.chunk_state.at(chunk_index) != CreateChunkState::REFERENCED) {
     RAY_LOG(DEBUG) << "Object " << object_id << " aborted due to OOM before chunk "
                    << chunk_index << " could be sealed";
     return;
   }
-  it->second.chunk_state[chunk_index] = CreateChunkState::SEALED;
```
Comment (rkooo567) [inline src/ray/object_manager/object_buffer_pool.cc:156]: Can you add a check here just in case (afaik, if exception is raised, it was not properly logged currently). 

```cpp
RAY_CHECK(it->second.chunk_info.size() >= chunk_index);
```

File: src/ray/object_manager/object_buffer_pool.h [hunk 2]
```diff
@@ -117,15 +117,17 @@ class ObjectBufferPool {
   /// \param chunk_index The index of the chunk.
   void AbortCreateChunk(const ObjectID &object_id, uint64_t chunk_index);
 
-  /// Seal the object associated with a create operation. This is invoked whenever
-  /// a chunk is successfully written to.
+  /// Write to a Chunk of an object.
```
Comment (rkooo567) [inline src/ray/object_manager/object_buffer_pool.h:120]: Can you also mention Seal is happening within this method? I feel like since Seal is an exposed concept to the upper layer, hiding it can confuse readers

**Review by rkooo567 — COMMENTED** [review]
Btw, can you tell me exactly which scenario can cause this? In a glance, it looks like the grpc server is multi threaded, but the handlePush is only happening in a single thread (rpc_service_), and Abort also seems to only happen on the same thread. 

Are you thinking the dependency manager calling CancelPull is the root cause? (it looks like the node manager and object manager uses the separate thread?), so it is like

- HandlePush is memcpying
- Dependency manager calls cancelPull before the memcpy is done
- And this cause problems

?

**Comment by scv119** [comment]
Following are an example of two threads competing on the same object, from the raylet who segfaulted.
thread 4581 created object and write the chunk, but figured it has been aborted when sealing it.
thread 4550 aborted the object creation while thread 4581 is writing to that object.

```
[2021-07-20 21:41:01,263 D 4550 4581] object_buffer_pool.cc:108: Created object 37c654c9b8db2110ffffffffffffffffffffffff0200000023000000 in plasma store, number of chunks: 1, chunk index: 0
[2021-07-20 21:41:01,265 D 4550 4550] pull_manager.cc:188: Deactivating pull for object 37c654c9b8db2110ffffffffffffffffffffffff0200000023000000
[2021-07-20 21:41:01,265 I 4550 4550] object_buffer_pool.cc:171: Not enough memory to create requested object 37c654c9b8db2110ffffffffffffffffffffffff0200000023000000, aborting
[2021-07-20 21:41:01,267 D 4550 4550] pull_manager.cc:394: Removing an object pull request of id: 37c654c9b8db2110ffffffffffffffffffffffff0200000023000000
[2021-07-20 21:41:01,267 D 4550 4581] object_buffer_pool.cc:152: Object 37c654c9b8db2110ffffffffffffffffffffffff0200000023000000 aborted due to OOM before chunk 0 could be sealed
```

File: src/ray/object_manager/object_manager.cc [hunk 3]
```diff
@@ -788,8 +788,7 @@ bool ObjectManager::ReceiveObjectChunk(const NodeID &node_id, const ObjectID &ob
   ObjectBufferPool::ChunkInfo chunk_info = chunk_status.first;
   if (chunk_status.second.ok()) {
     // Avoid handling this chunk if it's already being handled by another process.
-    std::memcpy(chunk_info.data, data.data(), chunk_info.buffer_length);
```
Comment (ericl) [inline src/ray/object_manager/object_manager.cc:791]: So chunk_info is holding a shared_ptr to the underlying PlasmaBuffer, so it seems that the memcpy should be safe as long as the chunk_info above is in scope.

I'm wondering if this is actually a new bug introduced by this fix: https://github.com/ray-project/ray/pull/16620/files

Here's my understanding:
- There are two refs to the underlying allocation, one from the PlasmaBuffer itself (decremented on destruction), and another from the client.
- AbortCreate() is (1) calling Release(), then (2) Abort(), then create_buffer_state.erase(object_id).
- Prior to 16620, by step (2) the ref-count drops to zero and plasma server can delete the object safely.
- But after 16620, there may be remaining shared_ptr refs to the PlasmaBuffer for the receive. Hence, we could get a Release() call for an already-deleted object.

If the above is correct, then this fix in this PR may be incomplete: you probably also want to remove the shared_ptr from ChunkInfo.

Comment (ericl) [reply src/ray/object_manager/object_manager.cc:791]: Aside: it's not clear to me why we need this special case Abort() path in the first place... ideally lifetime is managed only via the shared_ptr, and we don't have this separate dual reference counting path.

But maybe this is an issue for the refactor.

Comment (scv119) [reply src/ray/object_manager/object_manager.cc:791]: yup, I agree with your understanding is correct. Here is another race condition in the segfaulted raylet.  The following log validates the theory, as even after we released the object and then aborted the object,  there are still references to the object. (which presumably ref by the PlasmaBuffer from the chunkinfo).
```
[2021-07-20 21:33:31,506 D 4550 4575] store.cc:645: Erasing object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000 with nonzero ref count0b9920f3ca70541dffffffffffffffffffffffff010000004a000000, num bytes in use is now 1524494562
```



```
[2021-07-20 21:33:31,491 D 4550 4582] object_manager.cc:765: ReceiveObjectChunk on 228b63e21bcee0a7436682edf2de1fe6d7164b688467b094f3dc05d8 from 5c16b78c1611b84df7f908a9e718b0925121383f3a853c1424b5e037 of object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000 chunk index: 0, chunk data size: 5242880, object size: 45940869
[2021-07-20 21:33:31,506 D 4550 4550] pull_manager.cc:188: Deactivating pull for object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000
[2021-07-20 21:33:31,506 I 4550 4550] object_buffer_pool.cc:171: Not enough memory to create requested object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000, aborting
[2021-07-20 21:33:31,506 D 4550 4575] store.cc:633: Erasing object: 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000, address: 0x7f1eb9713cc0, size:45940869
[2021-07-20 21:33:31,506 D 4550 4575] store.cc:645: Erasing object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000 with nonzero ref count0b9920f3ca70541dffffffffffffffffffffffff010000004a000000, num bytes in use is now 1524494562
[2021-07-20 21:33:31,506 D 4550 4550] pull_manager.cc:394: Removing an object pull request of id: 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000
[2021-07-20 21:33:31,506 D 4550 4550] dependency_manager.cc:27: Object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000 no longer needed
[2021-07-20 21:33:31,507 D 4550 4583] object_buffer_pool.cc:152: Object 0b9920f3ca70541dffffffffffffffffffffffff010000004a000000 aborted due to OOM before chunk 5 could be sealed
```

Comment (ericl) [reply src/ray/object_manager/object_manager.cc:791]: Brainstorming another way to address this problem, perhaps we could have a Buffer state object that is responsible for calling Abort() on destruction? Then, Abort can be integrated into the normal C++ lifecycle management, and we can also avoid holding the mutex while writing to the buffer.

**Comment by scv119** [comment]
Yup I think that should work.

**Comment by rkooo567** [comment]
I think in the refactoring process, we should really make the lifecycle of this plasma object clear to avoid this sort of regression (and I believe it should be already in the scope) and avoid having many ways to delete objects like now ( like Abort / Delete )

**Comment by scv119** [comment]
address comments: we no longer holds an additional reference of PlasmaBuffer

**Review by ericl — APPROVED** [review]
LGTM. Probably can also remove the shared_ptr in ChunkInfo (should be un-used now) / move ChunkInfo to private.
